src/compiler/checker.ts

```diff
@@ -32029,6 +32029,59 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         );
     }
 
+    function discriminateContextualTypeByArrayElements(node: ArrayLiteralExpression, contextualType: UnionType) {
+        const key = `D${getNodeId(node)},${getTypeId(contextualType)}`;
+        const cachedType = getCachedType(key);
+        if (cachedType) return cachedType;
+
+        const isIndexPropertyLike = (property: Symbol) => /^[-+]?\d+$/.test(property.escapedName as string);
+        const hasIndexPropertyOutOfRange = (type: Type, length: number) =>
+            some(getPropertiesOfType(type), p => {
+                if (p.flags & SymbolFlags.Optional || !isIndexPropertyLike(p)) return false;
+                const index = +p.escapedName;
+                return index < 0 || index >= length;
+            });
+        const hasIndexPropertyInRange = (type: Type, length: number) =>
+            some(getPropertiesOfType(type), p => {
+                if (!isIndexPropertyLike(p)) return false;
+                const index = +p.escapedName;
+                return index >= 0 || index < length;
+            });

```

Comment:
I thiiink those are mostly redundant. An algorithm matching closer `discriminateContextualTypeByObjectMembers` would likely suffice here. You could create a list of potential discriminators based on the leading non-spread elements and pass that to `discriminateTypeByDiscriminableItems`. Then you might apply one extra filter after it (or before it?) to handle tuple-specific cases based on `.fixedLength` and similar